### ISSUE TITLE
Write a plan for a locking system to SOLVE ALL OUR LOCKING PROBLEMS

### ISSUE THREAD

**Issue by alaric-dotmesh** [dotmesh-io/dotmesh#707]
As a dev team, we want a way to "lock" dotmesh filesystems so we can easily solve our very many problems caused by concurrent operations on them.

## Relevant issues

* https://github.com/dotmesh-io/jupyterlab-plugin/issues/23

## Technical notes

We need some kind of application-level lock on a filesystem. It's fine for it to be an advisory lock, and a lot simpler to implement: we don't need to check lock ownership on dotmesh API operations. But it might constrain Dotmesh operations such as migrating the dot to another node, and therefore it interacts somewhat with the Procure call: does an app that procures a filesystem want to automatically get a lock to prevent it from being migrated to another node, for instance? I presume so, as if it's procuring it, then it probably expects the returned mount path to be valid at some unspecified point in the future - and it might not be currently...

**Comment by rusenask** [comment]
I would suggest to try investigating whether we could perform merges on a separate commit instead of using master branch tip, since when you open a PR it's not clear that your current tip is half merged. Also, since roll-back doesn't work, users can end up in a weird state. 

If we use a separate "staging" commit to do the merge, there would be no need to do a roll-back, we would just delete that merge commit.

**Comment by alaric-dotmesh** [comment]
Notes from a discussion about task locking:

When we start a task using a dot, we need to lock that dot on the hub to prevent uploads/merges/other tasks/whatever happening to that dot at the same time.

We can do that from the GW, as it knows the task state: get the lock when the task starts, release it when the task ends (be sure to track the actual state not the desired state, so that the lock is released when the agent process terminates after the pushing is all done, rather than when we ASK the agent to stop).

**Comment by itamarst** [comment]
Looking at the three categories above, seems like the optimization of having extra lock levels may not be worth it,and read/write should suffice?

**Comment by alaric-dotmesh** [comment]
Yes! Just read/write. The lock-against-migration-in-cluster can be a separate thing, as I think it's kind of orthogonal (that's the one I'll write a new issue for)

**Comment by alaric-dotmesh** [comment]
Rather, update #34 for as that's precisely what that issue is about :-)

**Comment by itamarst** [comment]
Was specifically thinking about "Reading the directory listing or files from the current filesystem state" which seems in-between the other two categories of operations in terms of restrictions. But also sounds bounded in time, so read/write should do the trick.

**Comment by alaric-dotmesh** [comment]
Ah yes, that can get a read lock, while the third category of "Operations that are probably safe to overlap with each other and with the above" just doesn't need to bother getting a lock

**Comment by itamarst** [comment]
Ahhhh, yes.

I am also going to do at least a quick think about whether actual locking (vs. advisory locking) might work.

**Comment by alaric-dotmesh** [comment]
Yeah! Mandatory locks would be an extra layer of protection against us making a mistake, so it would be great if we could do that too.

**Comment by itamarst** [comment]
I guess the main requirement for mandatory locks being sufficient is that you don't need multi-operation locks... and I'm getting impression there are Dotscience operations that require multiple Dotmesh operations.

**Comment by rusenask** [comment]
yeah, that's another problem :) dotmesh wasn't built for dotscience, we just started using it in a state that it was at that moment in time.

It would be good if:
a) New locking API was implemented using GRPC as that would give us an easy way to generate client code to use from the gateway
b) Built with authentication in mind from the start :)
c) Package should be a separate, not in the `main` package. I have started moving everything out but the RPC API is still there. If we get a new one in and then expand it to cover dotscience needs with one RPC call instead of multiple when we are doing something, then we could just safely remove current JSON RPC.

Auth on current JSON RPC was just bolted on and it JSON RPC also doesn't return any errors if functions fail to register, clients just get 404 which is very unhelpful error.

**Comment by itamarst** [comment]
Starting by looking at storage mechanism for locks. Given etcd is available, and designed for this sort of thing, looks like I want to add another etcd store to the config/inMemoryState/loading code in getKVDBStores, and use that?

**Comment by rusenask** [comment]
Locks should be exported via one of these: https://github.com/dotmesh-io/dotmesh/blob/master/pkg/store/store.go

as the https://github.com/dotmesh-io/dotmesh/blob/master/pkg/store/kvdb.go#L16 gives locking API that will work with etcd, consul, inmemory, boltdb implementations.

Regarding which store should expose locking, it's probably `FilesystemStore` as it has the filesystem records or maybe `RegistryStore`? not sure :) Another option is to create another interface that's separate from them.

**Comment by itamarst** [comment]
So going through repos that might need to be updated with locks, seems like:

* gateway
* dotscience-committer
* dotmesh-backup
* dotscience-agent

`ds` doesn't seem to use API directly, but the gateway might need to expose locking API for `ds`, maybe?

**Comment by itamarst** [comment]
The problem with advisory locks is that _everything_ doing a write operation needs to use them. And it seems like some of the code talks to DM via gateway, and some talks to it directly?

**Comment by itamarst** [comment]
> b) Built with authentication in mind from the start :)
> c) Package should be a separate, not in the `main` package. I have started moving everything out but the RPC API is still there. If we get a new one in and then expand it to cover dotscience needs with one RPC call instead of multiple when we are doing something, then we could just safely remove current JSON RPC.

Embedded in this comment (and some of the items listed as needing locks) is the notion that the units of operation in dotmesh are too low-level: there are higher-level dotscience "transactions" which are I suspect the units of locking, and it sounds like the units of authentication.

So another approach (or perhaps complementary approach) to advisory locking is switching all relevant dotscience clients away from talking to dotmesh directly, and only have them talk to ... the gateway? And the gateway can model the higher-level units of operation and ensure everything gets locking.

**Comment by itamarst** [comment]
And if we go down that route possibly the locking is something that is implemented in the gateway, not dotmesh. Though dotmesh should probably have its own mandatory locking, separately, it sounds like.

**Comment by rusenask** [comment]
I think ideally locking would just happen in the dotmesh but unfortunately we have scenarios where:

1. Gateway knows that the dot (volume) is being used by a runner somewhere else and no uploads should be allowed to a specific dot. Dotmesh doesn't know about it directly but if we allow to upload, we would get divergence. 
2. Gateway knows that the runner is not using dot anymore but the upload might be in progress, so probably dotmesh can now about it but there's no internal locking and cannot prevent uploads.

**Comment by alaric-dotmesh** [comment]
As an example of "Dotscience operations that require multiple Dotmesh operations", the process of merging a pull request from one filesystem to another is:

1. (DM API) Make a snapshot on the target filesystem saying "This is the start of a merge" with no actual filesystem changes (just snapshot metadata)
2. (DM API) Procure the target filesystem.
3. For each snapshot on the source filesystem that we want to pull in:
    1. (DM API) Mount that snapshot of the source filesystem.
    2. (Filesystem) Make the filesystem state of the target filesystem equal to the filesystem state of that snapshot in the source (currently, we're pretty crude about this and just copy the files over, but we want to make that more efficient to make better use of ZFS' copy-on-write snapshots)
    3. (DM API) Make a new snapshot in the target filesystem with the metadata of the source snapshot, plus some extra metadata to say it was pulled over
3. (DM API) Make a final metadata-only snapshot recording the end of the merge.

This involves a whole bunch of DM operations, and filesystem operations in the mounted filesystems; and if another operation kicks off while this is happening, it will both see a weird partially-merged state of the filesystem, and any changes it makes will mess up the snapshots that are supposed to be being copied verbatim form the source filesystem!

**Comment by itamarst** [comment]
Where locks are stored is somewhat orthogonal to who enforces them. In a world where only the gateway enforces locking:

1. All dotscience communication to dotmesh must go through gateway.
2. Gateway does the locking transparently to its clients; clients can't and don't request locks, the gateway does it for them.
3. Therefore, gateway must expose high-level operations.... although I guess in some cases it's "start operation" /"end operation" pairs, so there's requirement for liveness updates from clients to the gateway.

This ensures all operations are locked.

**Comment by itamarst** [comment]
Random thoughts on implementing locking: it can actually be mandatory locking and still support dotscience multi-operation use cases.

* Dotmesh gets new RPC calls: NewSession (returns a session ID), FinishSession.
* All RPC operations can optionally get a session ID as an HTTP header (available in the Request). If no session ID is given, a new session is generated, and it finishes at end of RPC operation.
* Locks conflict on session boundaries.
* Locks are held for scope of transaction.

This would involve adding locking to every RPC operation though.

**Comment by itamarst** [comment]
Another variant:

* Dotmesh gets Lock operation for a filesystem (returns a session ID) and Unlock.
* Session IDs get passed to all subsequent operations. Initially they do nothing (except maybe renew the lock lease?).
* Later we can add locking to each RPC operation in Dotmesh, incrementally, with reentrancy if it's relocking in an existing session.

**Comment by itamarst** [comment]
After conversation with @alaric-dotmesh , we concluded:

1. The gateway controls (at a high-level) all interactions with a particular fileysystem, there isn't any other process that somehow bypasses it for writes. If there are writes, the gateway initiated the process.
2. Dotmesh on its own might need locking for other reasons, but for now that's a smaller scope and less important.
3. Therefore:
    * Locking should be done transparently by the gateway, as part of the high-level operations it models.
    * Lock state can be stored in the gateway database.
